BiometricsFingerprint.cpp

```diff
@@ -157,68 +169,83 @@ Return<RequestStatus> BiometricsFingerprint::enumerate() {
+    if (!rc) {
+        if (!print_indices.print_count)
+            // When there are no fingers, the service still needs to know that (potentially async)
+            // enumeration has finished. By convention, send fid=0 and remaining=0 to signal this:
+            mClientCallback->onEnumerate(devId, 0, mDevice->gid, 0);
+        else
+            for (size_t i = 0; i < print_indices.print_count; i++) {
+                ALOGD("%s : found print : %lu at index %zu", __func__, (unsigned long)print_indices.prints[i], i);
 
-            uint32_t  remaining_templates = (uint32_t)(print_indexs.print_count - i - 1);
+                uint32_t remaining_templates = (uint32_t)(print_indices.print_count - i - 1);
 
-            mClientCallback->onEnumerate(devId, print_indexs.prints[i], mDevice->gid, remaining_templates);
-        }
+                mClientCallback->onEnumerate(devId, print_indices.prints[i], mDevice->gid, remaining_templates);
+            }
+    }
 
     mWt.Resume();
 
-    return ErrorFilter(0);
+    return ErrorFilter(rc);
 }
 
 Return<RequestStatus> BiometricsFingerprint::remove(uint32_t gid, uint32_t fid) {
-
     const uint64_t devId = reinterpret_cast<uint64_t>(mDevice);
 
     if (mClientCallback == nullptr) {
         ALOGE("Client callback not set");
-        return ErrorFilter(-1);
+        return RequestStatus::SYS_EINVAL;
     }
 
     if (!mWt.Pause())
         return RequestStatus::SYS_EBUSY;
 
-    Return<RequestStatus> ret = RequestStatus::SYS_OK;
+    int rc = 0;
 
-    if (fpc_del_print_id(mDevice->fpc, fid) == 0){
+    if (fid == 0) {
+        // Delete all fingerprints when fid is zero:
+        ALOGD("Deleting all fingerprints for gid %d", gid);
 
-        mClientCallback->onRemoved(devId, fid, gid,0);
+        fpc_fingerprint_index_t print_indices;
+        rc = fpc_get_print_index(mDevice->fpc, &print_indices);
+        if (!rc)
+            for (auto remaining = print_indices.print_count; remaining--;) {
+                auto fid = print_indices.prints[remaining];
+                ALOGD("Deleting print %d, %d remaining", fid, remaining);
+                rc = fpc_del_print_id(mDevice->fpc, fid);
+                if (rc)
+                    break;
+                mClientCallback->onRemoved(devId, fid, gid, remaining);
+            }
+    } else {
+        ALOGD("Removing finger %u for gid %u", fid, gid);
+        rc = fpc_del_print_id(mDevice->fpc, fid);
+        if (!rc)
+            mClientCallback->onRemoved(devId, fid, gid, 0);
+    }
 
-        uint32_t db_length = fpc_get_user_db_length(mDevice->fpc);
-        ALOGD("%s : User Database Length Is : %lu", __func__,(unsigned long) db_length);
-        fpc_store_user_db(mDevice->fpc, db_length, mDevice->db_path);
-        ret = ErrorFilter(0);
+    if (rc) {
+        mClientCallback->onError(devId, FingerprintError::ERROR_UNABLE_TO_REMOVE, 0);
     } else {
-        mClientCallback->onError(devId, FingerprintError::ERROR_UNABLE_TO_REMOVE, -1);
```

Comment:
You changed the last argument from -1 to 0. I presume you know what you are doing; flagging here just in case of a typo.

Reply:
That's a good catch, didn't expect anyone noticing it :)

The other invocations in this file (and Egistec) use `0`, for a "good" reason. Android uses these error codes to look up a vendor error string in an array. Use `-1` and you get a nice out-of-bounds exception in the framework.
That said this change is only to prevent our future selves from messing up: The error code is _not_ `FingerprintError::ERROR_VENDOR` meaning this vendorCode is completely ignored. In the event that we do start using it and copy/modify this line without updating the vendor code we're safe.

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/hardware/fingerprint/FingerprintManager.java;l=971-973;drc=7c7d90b8c6f31039242ae95faac837fd69a373f4

Reply:
LGTM; thank you!